member-processor.ts

```diff
@@ -0,0 +1,105 @@
+import { Writable } from 'stream';
+import * as RDF from "rdf-js";
+import { TreeProperties, extractVersionTimestamp, extractBaseResourceUri, convertBlankNodes } from "./utils";
+import { LDES_VERSION_OF_PATH, LDES_TIMESTAMP_PATH, REPLACE_VERSIONS } from "./config";
+import { executeDeleteInsertQuery, getLatestTimestamp } from "./sparql-queries";
+import { DataFactory } from "n3";
+const { quad, variable } = DataFactory;
+
+
+export type Member = {
+  id: RDF.Term;
+  quads: RDF.Quad[];
+};
+
+type  MemberWithCallBack = {
+  member: Member;
+  callback: (e?: Error) => void;
+}
+
+function timeout (ms: number) {
+  return new Promise(resolve => {
+    setTimeout(resolve, ms);
+  });
+}
+
+export default class MemberProcessor extends Writable {
+  private treeProperties: TreeProperties;
+  private latestVersionMap : Map<string, Date> = new Map();
+  membersToProcess: MemberWithCallBack[] = [];
+
+  constructor() {
+    super({ objectMode: true, highWaterMark: 1000 });
+    this.treeProperties = {
+      versionOfPath: LDES_VERSION_OF_PATH,
+      timestampPath: LDES_TIMESTAMP_PATH
+    };
+    this.processingLoop();
+  }
+
+  _write(member: Member, _encoding : string , callback: () => void) {
+    this.membersToProcess.push({member, callback});
+    return true;
+  }
+
+  async processingLoop () {
+    do {
+      const next = this.membersToProcess.shift();
+      if (next) {
+        try {
+          await this.processMember(next.member);
+          await next.callback();
+        }
+        catch (e) {
+          console.error(e);
+          await next.callback(e);
+          this.destroy(e);
+        }
+      }
+      await timeout(10);
+    }  while (! this.closed);
```

Comment:
Not 100% sure, but I think this might be an infinite loop since `this.closed` doesn't seem to be set to true anywhere?